### ISSUE TITLE
Could turbo-frame elements dispatch turbo events during their navigation lifecycle?

### ISSUE THREAD

**Issue by seanpdoyle** [hotwired/turbo#54]
As discussed in [Are transitions and animations on Hotwire roadmap?](https://discuss.hotwire.dev/t/are-transitions-and-animations-on-hotwire-roadmap/1547), @sstephenson mentions that there would be utility in (exposing appropriate seams to consumer applications](https://discuss.hotwire.dev/t/are-transitions-and-animations-on-hotwire-roadmap/1547/4).

One potential solution could involve dispatching the existing [`turbo:`-prefixed events](https://turbo.hotwire.dev/reference/events) as a `<turbo-frame>` element progresses through a navigation. Those events could be dispatched and _bubble up from_ the `turbo-frame` element, making the frame available as the `CustomEvent.target` when appropriate. Page-level `turbo:`-prefixed events would still dispatch off the `document` element.

In addition to the value it would bring for ancestors to be notified of various lifecycle hooks, it could also enable the cancelation of `turbo-frame` visitations, and could expose timing metrics as part of frame-scoped `turbo:load` events.

From a technical perspective, the events could be dispatched from the various empty `FrameController` hooks:

https://github.com/hotwired/turbo/blob/c52d87c9426ebbbb735e9e8947b3cbdfde8a2a10/src/core/frames/frame_controller.ts#L75-L121

Alternatively,  a `FrameController` could be provided with its own [`Session` instance](https://github.com/hotwired/turbo/blob/c52d87c9426ebbbb735e9e8947b3cbdfde8a2a10/src/core/session.ts#L20) scoped to the `<turbo-frame>` element (as opposed to the current implementation which operates on the `document.documentElement` element).

In theory, there could be a lot of value in unifying the implementations of how a page and a frame navigate, but might involve a larger set of architectural changes. Are there conceptual mismatches between a frame-scoped [`Session`](https://github.com/hotwired/turbo/blob/c52d87c9426ebbbb735e9e8947b3cbdfde8a2a10/src/core/session.ts) / [`Navigator`](https://github.com/hotwired/turbo/blob/c52d87c9426ebbbb735e9e8947b3cbdfde8a2a10/src/core/drive/navigator.ts) pairing and the current implementation which relies on [`FormInterceptor`](https://github.com/hotwired/turbo/blob/c52d87c9426ebbbb735e9e8947b3cbdfde8a2a10/src/core/frames/form_interceptor.ts) and [`LinkInterceptor`](https://github.com/hotwired/turbo/blob/c52d87c9426ebbbb735e9e8947b3cbdfde8a2a10/src/core/frames/link_interceptor.ts) instances?

**Comment by jonsgreen** [comment]
@seanpdoyle Yes, and now I see that hotwired/turbo would have been the right place to have looked for similar issues before posting my issue. We can close this issue though I am still curious about the best way to get involved with any of the hotwire projects.
